ODIN_II/SRC/HardSoftLogicMixer.cpp

```diff
@@ -0,0 +1,237 @@
+/*
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following
+ * conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "HardSoftLogicMixer.hpp"
+#include <iostream>
+#include <vector>
+#include <cstdint>
+#include <cstring>
+#include "OdinGridAnalyzer.hpp"
+#include <bits/stdc++.h>
+#include "partial_map.h" // instantiate add_with_carry
+#include "adders.h"      // instantiate_hard_adders
+
+HardSoftLogicMixer::HardSoftLogicMixer(t_arch& arch, const config_t config, std::vector<t_physical_tile_type> tileTypes)
+    : _analyzer()
+    , _arch(arch) {
+    _allOptsDisabled = true;
+    _tileTypes = tileTypes;
+    // By default, disables all optimizations
+    for (int i = 0; i < mix_hard_blocks::Count; i++) {
+        _enabledOptimizations[i] = false;
+    }
+
+    parseAndSetOptimizationParameters(config);
+}
+
+void HardSoftLogicMixer::parseAndSetOptimizationParameters(const config_t config) {
+    if (config.mix_soft_and_hard_logic != 0) {
+        int checkValue = -1;
+        int mix_soft_and_hard_logic = config.mix_soft_and_hard_logic;
+        int temp;
+        for (int i = 0; i < mix_hard_blocks::Count; i++) {
+            temp = 1 << (i);
+            checkValue = temp & mix_soft_and_hard_logic;
+            if (checkValue != 0) {
+                _enabledOptimizations[i] = true;
+            }
+        }
+        _allOptsDisabled = false;
+        //Explicitly set all hard block numbers to max
+        for (int i = 0; i < mix_hard_blocks::Count; i++) {
+            _hardBlocksCount[i] = INT_MAX;
+            _hardBlocksMixingRatio[i] = 1.0f;
+        }
+
+        _hardBlocksMixingRatio[mix_hard_blocks::MULTIPLIERS] = config.mults_mixing_ratio;
```

Comment:
is there any way to make this generic and not just stuck on Multipliers?

Reply:
Yes, we can work on that for sure.

Reply:
@jeanlego, I've decided to commit just the multipliers first. This PR has the minimum required code to allow new functionality. Hence, this is left as a future work.

Reply:
it is okay by me, just make sure to leave traces of how you would do it in here for the next person (could be you in a year! you never know :P)

Reply:
I like hows you've done it, I would maybe prefer if you hooked onto the enum in odin_types.h to make it easier (since our harblocks are registered in that enum anywho)

Reply:
https://github.com/CAS-Atlantic/vtr-verilog-to-routing/pull/7/files#diff-f65d2282d9b36c0c92239bc89434c7a5R51-R55 sort of address that,but it might be not enough, I agree. But I also plan to continue working on the optimization after (if) it gets in.